pandas/core/indexing.py

```diff
@@ -1936,42 +1935,31 @@ def _setitem_single_column(self, loc: int, value, plane_indexer):
         """
         pi = plane_indexer
 
-        ser = self.obj._ixs(loc, axis=1)
-
         # perform the equivalent of a setitem on the info axis
         # as we have a null slice or a slice with full bounds
         # which means essentially reassign to the columns of a
         # multi-dim object
         # GH#6149 (null slice), GH#10408 (full bounds)
         if com.is_null_slice(pi) or com.is_full_slice(pi, len(self.obj)):
-            ser = value
+            self.obj._iset_item(loc, value)
         elif (
             is_array_like(value)
-            and is_exact_shape_match(ser, value)
+            and len(value.shape) > 0
+            and self.obj.shape[0] == value.shape[0]
             and not is_empty_indexer(pi)
         ):
             if is_list_like(pi):
-                ser = value[np.argsort(pi)]
+                value = value[np.argsort(pi)]
             else:
                 # in case of slice
-                ser = value[pi]
+                value = value[pi]
+            self.obj._iset_item(loc, value)
         else:
             # set the item, first attempting to operate inplace, then
```

Comment:
is this comment still accurate/helpful?

Reply:
Hmm, not really I think (but also not before this PR). The logic of "first attempt to operate inplace, then falling back to casting if necessary" is contained within the `column_setitem` call (and before this PR it was contained in the `setitem` call)

Reply:
Now, it was added in https://github.com/pandas-dev/pandas/commit/573c063c67b2433efb2210a74065e4406032eb0d, and from that point of view: I would say the comment is still as accurate as it was when it was added :)

But I think we can also remove it now, it's the general behaviour of loc/iloc to fallback to casting

Reply:
Might be worth discussing the gameplan for after the deprecation inside _setitem_single_column is enforced.  What I had in mind was that inside of doing the can_hold_element check as part of take_split_path, we would do a try/except falling back to the split path.  Does that idea play nicely with what you're doing here?